### PR TITLE
Fix queue len metric maths

### PR DESCRIPTION
**Pull request by nyanzebra**
After doing some testing, I have found this type of approach should work. We will pass into propose the total number of messages that store is aware of, this way we can adjust in the event of TTL.

Imagine there are 100 messages being processed and half of them expire leaving 50. The checkpointer would still have its' two pointers for propose and offset remain the same, showing a difference of 100 (100 - 0). But in reality, we should show 50. To do this we can say that if (proposed - offset) > total_messages, move offset closer to proposed. To accommodate this change the propose method for checkpointers has been updated.

Changes:
- Propose signature for checkpointers
- update columnFamilyDbStore for correct count and make count operations atomic
- use endpoint executor to pass info that MessageStore knows about to Checkpointer

### CODE REVIEW

File: edge-hub/core/src/Microsoft.Azure.Devices.Routing.Core/endpoints/StoringAsyncEndpointExecutor.cs [hunk 1]
```diff
@@ -251,6 +266,14 @@ async Task SendMessagesPump()
                                 await this.ProcessMessages(messages, fsm);
                                 Events.SendMessagesSuccess(this, messages, priority);
 
+                                // Update queue len metric
+                                var count = Math.Max(await this.messageStore.GetMessageCount(this.Endpoint.Id) - (ulong)messages.Length, 0);
+                                if (this.messageStoreQueueLengths.TryGetValue(this.Endpoint.Id, out var queueLength))
+                                {
+                                    count = queueLength;
```
Comment (varunpuranik) [inline edge-hub/core/src/Microsoft.Azure.Devices.Routing.Core/endpoints/StoringAsyncEndpointExecutor.cs:273]: Why do you need to do this?

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/storage/MessageStore.cs [hunk 2]
```diff
@@ -360,8 +373,6 @@ async Task<bool> DeleteMessageCallback(long offset, MessageRef messageRef)
                                 }
                             }
 
-                            // update Metrics for message counts
```
Comment (varunpuranik) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/storage/MessageStore.cs:363]: How does cleanup get accounted for, if you remove this code?

**Comment by nyanzebra** [comment]
Accidentally closed

File: edge-hub/core/src/Microsoft.Azure.Devices.Routing.Core/endpoints/StoringAsyncEndpointExecutor.cs (hunk 1 above)
Comment (nyanzebra) [reply edge-hub/core/src/Microsoft.Azure.Devices.Routing.Core/endpoints/StoringAsyncEndpointExecutor.cs]: Not needed

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/storage/MessageStore.cs (hunk 2 above)
Comment (nyanzebra) [reply edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/storage/MessageStore.cs:363]: Cleanup setting of count will happen

**Comment by nyanzebra** [comment]
I am trying to see the cleanup manual test working... however, the cleanup task starts and never finishes (or is going to take time). I am testing on my pi4b and will wait to see if it completes. There shouldn't be any changes in the pr that would cause excessive waiting but maybe I have a local problem...

**Comment by nyanzebra** [comment]
> I am trying to see the cleanup manual test working... however, the cleanup task starts and never finishes (or is going to take time). I am testing on my pi4b and will wait to see if it completes. There shouldn't be any changes in the pr that would cause excessive waiting but maybe I have a local problem...

Once I see cleanup complete and a satisfactory graph then I will post here and we can hopefully close this pr

**Comment by nyanzebra** [comment]
image
As can be seen from this graph we are correctly growing and returning to 0 after going from online -> offline -> online. The one thing that seems weird is the initial cleanupqueue call at EH start does not seem to execute in a timely manner... this leads the graph not picking up where it left off and instead restarting from 0... This change shouldn't have anything to do cleanup queue really, so this is odd behaviour to me...

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/storage/MessageStore.cs [hunk 3]
```diff
@@ -55,31 +55,31 @@ public void SetTimeToLive(TimeSpan timeSpan)
             Events.TtlUpdated(timeSpan);
         }
 
-        public async Task AddEndpoint(string endpointId)
+        public async Task AddEndpoint(string messageQueueId)
```
Comment (nyanzebra) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/storage/MessageStore.cs:58]: Updated these based on what I saw its' usage really is from StoringEndpoint

**Comment by nyanzebra** [comment]
> image
> As can be seen from this graph we are correctly growing and returning to 0 after going from online -> offline -> online. The one thing that seems weird is the initial cleanupqueue call at EH start does not seem to execute in a timely manner... this leads the graph not picking up where it left off and instead restarting from 0... This change shouldn't have anything to do cleanup queue really, so this is odd behaviour to me...

I have confirmed correct behavior on both release and master branches. So this change is the problem for the cleanup not completing quickly. I know it is not completing because I grep'ed the logs for it see the start log message but no finish log message until after 30 min...

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/storage/MessageStore.cs (hunk 3 above)
Comment (varunpuranik) [reply edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/storage/MessageStore.cs:58]: This change makes sense.. but - 
1. this is old code, and I would rather not change it at this point
2. we use endpointId for this everywhere else in the code (that was originally what we were calling it). 
So I would suggest we keep it as it is to minimize the changes here..

Comment (nyanzebra) [reply edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/storage/MessageStore.cs:58]: Sounds good, I am okay with reverting back to endpointId for the variable as long as it is understood that it doesn't seem to always be the case (at least based on what I saw)

Comment (nyanzebra) [reply edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/storage/MessageStore.cs:58]: So, maybe a comment will suffice?

File: edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/storage/MessageStore.cs [hunk 4]
```diff
@@ -360,8 +362,11 @@ async Task<bool> DeleteMessageCallback(long offset, MessageRef messageRef)
                                 }
                             }
 
-                            // update Metrics for message counts
-                            Checkpointer.Metrics.SetQueueLength(await sequentialStore.Count(), endpointId, priority.ToString());
+                            // Since we will have the total *true* count of messages in the store, we need to set the counter here
+                            // to the total count of messages in the store.
+                            var count = await sequentialStore.Count();
+                            Checkpointer.Metrics.SetQueueLength(count, endpointId, priority);
```
Comment (nyanzebra) [inline edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Core/storage/MessageStore.cs:368]: I can only assume this is somehow causing the weird behaviour in the graph I linked with this pr where we do not pick up on queue length where we left off... both master and release/1.2 work as expected... and there aren't too many other places in this code that have changed...

File: edge-hub/core/src/Microsoft.Azure.Devices.Routing.Core/checkpointers/Checkpointer.cs [hunk 5]
```diff
@@ -111,7 +111,11 @@ public async Task CommitAsync(ICollection<IMessage> successful, ICollection<IMes
                 this.LastFailedRevivalTime = lastFailedRevivalTime;
                 this.UnhealthySince = unhealthySince;
                 await this.store.SetCheckpointDataAsync(this.Id, new CheckpointData(offset, this.LastFailedRevivalTime, this.UnhealthySince), token);
-                Metrics.SetQueueLength(this);
+            }
+
+            foreach (var message in successful)
+            {
+                Metrics.DecrementQueueLength(this.EndpointId, uint.Parse(this.Priority));
```
[the remark this answers is not in the document]
Comment (nyanzebra) [reply edge-hub/core/src/Microsoft.Azure.Devices.Routing.Core/checkpointers/Checkpointer.cs:118]: Also, is there some nice way of doing this? Like succesful.ForEach(my_fn)? After enough rust, I feel like this is the way *I* should write things lol

[the remark this answers is not in the document]
Comment (vadim-kovalyov) [reply edge-hub/core/src/Microsoft.Azure.Devices.Routing.Core/checkpointers/Checkpointer.cs]: I see what you mean :) There is `List<T>.ForEach()` that I'm aware of, but IMHO, `foreach` keyword is cleaner in this case.

For projections (aka `map`, aka `Select`) I would, of course, use extension methods from LINQ.

File: edge-hub/core/src/Microsoft.Azure.Devices.Routing.Core/checkpointers/Checkpointer.cs [hunk 6]
```diff
@@ -198,18 +202,28 @@ static string GetContextString(Checkpointer checkpointer)
 
         public static class Metrics
         {
+            static readonly ConcurrentDictionary<string, ulong> Counters = new ConcurrentDictionary<string, ulong>();
```
Comment (varunpuranik) [inline edge-hub/core/src/Microsoft.Azure.Devices.Routing.Core/checkpointers/Checkpointer.cs:205]: I wonder if this is the best way to do it, or if we should add an Increment/Decrement methods to IMetricsGauge.. 
I see that IMeasureGaugeMetrics has an overload that allows you to Get the metric value before setting it - 
https://github.com/AppMetrics/AppMetrics/blob/features/4.2.0/src/Core/src/App.Metrics.Abstractions/Gauge/IMeasureGaugeMetrics.cs#L30
Maybe we can use that? 

Another option could be IMeasureCounterMetrics, but it doesn't seem to have a SetValue method - https://github.com/AppMetrics/AppMetrics/blob/features/4.2.0/src/Core/src/App.Metrics.Abstractions/Counter/IMeasureCounterMetrics.cs

Comment (nyanzebra) [reply edge-hub/core/src/Microsoft.Azure.Devices.Routing.Core/checkpointers/Checkpointer.cs:205]: So, I think we want to use gauge. Counter, I think, is only for 'increasing' values like number of messages handled. Since we know that we need to set after cleanup and do decrements I am not sure Counter will work.

For the updating the interface, I think that is a great way of doing this, especially since the prometheus version of gauge already has these methods :)

File: edge-hub/core/src/Microsoft.Azure.Devices.Routing.Core/checkpointers/Checkpointer.cs [hunk 7]
```diff
@@ -35,6 +35,8 @@ public class Checkpointer : ICheckpointer
             this.closed = new AtomicBoolean(false);
             this.EndpointId = endpointId;
             this.Priority = priority;
+
+            Metrics.SetQueueLength((ulong)Math.Max(this.Offset, 0), endpointId, priority);
```
Comment (varunpuranik) [inline edge-hub/core/src/Microsoft.Azure.Devices.Routing.Core/checkpointers/Checkpointer.cs:39]: This does not seem correct. The Queue length would be Queue.TailPointer - Checkpointer.Offset.

Comment (nyanzebra) [reply edge-hub/core/src/Microsoft.Azure.Devices.Routing.Core/checkpointers/Checkpointer.cs:39]: Both the offset and proposed are set to the CheckpointData.Offset. This is also all there is on the checkpointData (Offset, LastFailedRevivalTime and UnhealthySince). Does this not map to what you would expect?

File: edge-hub/core/test/Microsoft.Azure.Devices.Routing.Core.Test/checkpointers/CheckpointerTest.cs [hunk 8]
```diff
@@ -203,35 +217,56 @@ public void TestQueueLengthMetricSetAndDecrement()
             t1.Join();
             t2.Join();
 
-            Assert.Equal(0u, Checkpointer.Metrics.GetQueueLengthCounter("endpoint", 0));
+            Assert.True(0u >= Checkpointer.Metrics.GetQueueLength("endpoint", 0));
         }
 
         [Fact]
         [Unit]
         public void TestQueueLengthMetricIncrement()
```
Comment (nyanzebra) [inline edge-hub/core/test/Microsoft.Azure.Devices.Routing.Core.Test/checkpointers/CheckpointerTest.cs]: Not sure why, but it seems like maybe in CI these tests don't actually use the real metrics provider but maybe are using the NullMetricsProvider... Any ideas on this? I am looking into test infrastructure now to find out if there is something different for CI than what I see locally.

File: edge-hub/core/src/Microsoft.Azure.Devices.Routing.Core/checkpointers/Checkpointer.cs (hunk 7 above)
Comment (nyanzebra) [reply edge-hub/core/src/Microsoft.Azure.Devices.Routing.Core/checkpointers/Checkpointer.cs:39]: This idea of using createAsync cannot work... It must be the case that if we have 10 messages all unprocessed... then the offset is 0. If this value was 10 then on propose we would get 11 - 10 for count of metric and in logs but that is not the case.

We also don't want to put in 'propose' call as that leads to same problem as we have now... let's discuss this a bit, as we need a new strategy. What might be the 'easiest' thing to do with very few changes... would to have checkpointer passed into MessageStore or have something exposed on Checkpointer to where we can take the count of messages from cleanup and 'adjust' the proposed and offset values. As the problem with current implementation is TTL where Checkpointer has no idea about messages that just disappear after 300 seconds (or whatever value). Still let's talk about this for bit and think this through :)

**Comment by nyanzebra** [comment]
image

Here we see that things are working as expected, EH is restarted and line picks up at same slope, then after 30 min it drops down to 60 from 300 something (as TTL is 300s and send rate is 5s => 60 max msgs in window for TTL). Then I switch to online again and we see line go back to 0.

File: edge-util/src/Microsoft.Azure.Devices.Edge.Storage.RocksDb/ColumnFamilyDbStore.cs [hunk 9]
```diff
@@ -12,14 +12,15 @@ namespace Microsoft.Azure.Devices.Edge.Storage.RocksDb
     class ColumnFamilyDbStore : IDbStore
     {
         readonly IRocksDb db;
-        private ulong count;
+        private long count;
```
Comment (varunpuranik) [inline edge-util/src/Microsoft.Azure.Devices.Edge.Storage.RocksDb/ColumnFamilyDbStore.cs:15]: why?

File: edge-hub/core/src/Microsoft.Azure.Devices.Routing.Core/checkpointers/Checkpointer.cs [hunk 10]
```diff
@@ -68,9 +68,20 @@ public static async Task<Checkpointer> CreateAsync(string id, ICheckpointStore s
...
         {
             this.Proposed = Math.Max(message.Offset, this.Proposed);
+
+            if (totalMessages.HasValue)
+            {
+                var messageCount = (ulong)(this.Proposed - this.Offset);
+                var count = totalMessages.GetOrElse(ulong.MaxValue);
+                if (messageCount > count)
+                {
+                    this.Offset = this.Proposed - (long)count;
```
Comment (varunpuranik) [inline edge-hub/core/src/Microsoft.Azure.Devices.Routing.Core/checkpointers/Checkpointer.cs:81]: What is the logic here? This seems unrelated to the metric.... Are you saying the offset is incorrect in EdgeHub?

Comment (nyanzebra) [reply edge-hub/core/src/Microsoft.Azure.Devices.Routing.Core/checkpointers/Checkpointer.cs:81]: This is what I mean by adjusting the offset. Without TTL these values should be exactly the same, but with TTL there can be a discrepancy.

Example 1: No TTL and we are processing normally
We have 100 messages proposed and say 20 processed. This means that the count should be 80. 30 minutes have not passed so the totalMessages are still 100. We want to show 80.

Example 2: No TTL and we are offline
Same as above, we should always show what proposed - offset equals

Example 3: TTL and we process normally
Say we have 100 messages and 20 have been processed. Let's pretend we hit the 30 min mark and remove 40 messages due to TTL. This means we really only have 60 messages and not 80. So, we should adjust offset to reflect this fact.

Example 4: TTL and offline
At 30 min mark for cleanup we should only have TTL * (number of messages / second). So, if TTL is 300 seconds and we send 1 message every 5 seconds then there can only be a max of 60 messages. So, if we have say 200 messages proposed and none processed; at the 30 min mark we should adjust the offset to reflect we only have 60 messages that can possibly be processed.

Example 5: Come back from crash
Much the same as above examples, after a crash we want to make sure we only count the messages that are actually there. Currently, the checkpointer takes the message offset as the way of determining how many messages. For details on this, if we crashed after processing 100 messages then came up and got a new message, that message would be marked as 101 and we would have 1 message left to process. If we instead had processed 200 and had an offset of 100, but due to TTL eliminated another 40 messages, then we shouldn't say there are 100 messages to process (queue length) but instead we should say 60.

**Review by vadim-kovalyov — DISMISSED** [review]
